last2libre/scrobblelib.py

```diff
@@ -0,0 +1,122 @@
+import json
+import time
+
+from urllib.parse import urlencode
+from urllib.request import Request, URLError, HTTPError, urlopen
+
+
+class ScrobbleException(Exception):
+    pass
+
+
+class InvalidScrobbleServer(Exception):
+    pass
+
+
+class ScrobbleServer(object):
+    def __init__(
+            self, server_url, session_key, api_key, debug=False,
+            username=False):
+        self.server_url = server_url
+        self.session_key = session_key
+        self.api_key = api_key
+        self.debug = debug
+
+        self.log = None
+        self.post_data = []
+        if debug:
+            self.log = open(username + '.response.log', 'w+')
+
+    def submit(self):
+        if len(self.post_data) == 0:
+            return
+
+        i = 0
+        data = []
+        last_error = None
+
+        for listen in self.post_data:
+            data += listen.get_tuples(i)
+            i += 1
+
+        data += [
+            ('method', 'track.scrobble'),
+            ('sk', self.session_key),
+            ('api_key', self.api_key),
+            ('format', 'json'),
+        ]
+
+        for timeout in (1, 2, 4, 8, 16, 32):
```

Comment:
What exactly do these numbers represent? It seems `timeout` is only ever called in the Exception. Is there any specific use, or can you use a normal loop from 0 to 5? It looks like these are 2^n seconds per "retry."

Reply:
On [L76](https://github.com/jwflory/last2libre/blob/414f6032fe41cdd3d99f8216d9f5796156ac0f1b/last2libre/scrobblelib.py#L76), the `timeout` interval is used in a `time.sleep()` call to rate-limit failed attempts to hit the API. There may be a more efficient way of doing this, but I'm not sure what it would be.